mountpoint-s3/src/inode.rs

```diff
@@ -382,77 +370,95 @@ impl SuperblockInner {
             .ok_or(InodeError::InodeDoesNotExist(ino))
     }
 
-    /// Find the inode with the given name in a parent directory and update its `stat`. If the inode
-    /// does not exist, create it if possible.
-    pub fn update_or_insert(
+    /// Update the inode with the given name in a parent directory with the remote data.
+    /// It may update or delete an existing inode, or insert a new one.
+    pub fn update_from_remote(
         &self,
         parent_ino: InodeNo,
         name: &str,
-        stat: InodeStat,
-        kind: InodeKind,
-    ) -> Result<Inode, InodeError> {
-        let parent = {
-            let inodes = self.inodes.read().unwrap();
-            inodes
-                .get(&parent_ino)
-                .cloned()
-                .ok_or(InodeError::InodeDoesNotExist(parent_ino))?
-        };
+        remote: Option<RemoteLookup>,
+    ) -> Result<LookedUp, InodeError> {
```

Comment:
I still have mixed feelings about this function, the name is `update_from_remote` but we're still updating it even if remote data is None and then we return `LookedUp` as a result. So, I'm not sure if we should have another layer to call this function when there is something to update, and also control when to remove (remove itself could be another function).

Reply:
I see your point. I would still prefer to keep all the logic in this function, but we definitely need a better name/description for it!

In my mind, this is where we take what we got from the client (`Option<RemoteLookup>`) and update our local state:
* if both inode and key exist - or neither do! - we use the fast path,
* otherwise, we need to write-lock the parent to insert or delete an inode.

Moving the deletion outside would mean duplicating the locking logic somewhere else.

Reply:
Yeah, I understand that you are trying to avoid multiple locking and cloning but it's not really straightforward when we read function/struct definition (e.g. why include `LookedUp` in `UpdateStatus::Updated` instead of `Inode` like we do in `UpdateStatus::LocalOnly`).

Let's keep the logic in one function and update the name/description to make it clearer.